src/mod/endpoints/mod_sofia/sofia_glue.c

```diff
@@ -3294,6 +3294,65 @@ void sofia_glue_build_vid_refresh_message(switch_core_session_t *session, const
 }
 
 
+char *sofia_glue_get_encoded_fs_path(nua_handle_t *nh, sip_route_t *rt, switch_bool_t add_fs_path_prefix)
+{
+	char *route;
+	int count = 0;
+	char *routes[ROUTE_MAX_HEADERS] = {0};
+	int buf_len_required = (add_fs_path_prefix ? 10 : 1);  /* ;fs_path=  + nul byte*/
+	int buf_len_avail;
+	char *route_encoded_buf = NULL;
+	char *buf_p = NULL;
+
+	for (sip_route_t *rr = rt; rr; rr = rr->r_next) {
```

Comment:
Can we use `SWITCH_STANDARD_STREAM` here to avoid having multiple loops, `routes` array, the need for calculating `buf_len_required` and calling `switch_zmalloc` ?

Reply:
> 

If you feel that is better. Ok. 
Internally SWITCH_STANDARD_STREAM may need to call multiple times malloc since you don't know the length before hand.
I thought, if you can do one malloc, why not.
Also, you still need a buffer of known length for switch_url_encode

Reply:
Let's try moving to the stream. 
Something like this?
Here is an example (not a draft or a real implementation) for you:
```
		sip_record_route_t *rrp;
		switch_stream_handle_t rr_stream = { 0 };
		int count = 0;
		SWITCH_STANDARD_STREAM(rr_stream);
		for(rrp = sip->sip_record_route; rrp; rrp = rrp->r_next) {
			char *rr = sofia_glue_strip_uri(sip_header_as_string(nua_handle_home(nh), (void *) rrp));
			rr_stream.write_function(&rr_stream, count == 0 ? "%s" : ",%s", rr);
			free(rr);
			if (count == 127) break;
			count++;
		}
		if (!zstr((char *) rr_stream.data)) {
			route = rr_stream.data;
		} else {
			route = sofia_glue_strip_uri(sip_header_as_string(nua_handle_home(nh), (void *) sip->sip_record_route));
			switch_safe_free(rr_stream.data);
		}
```